[BUG] Build Id mismatch while running `go test` using `-covermode` and `-coverprofile`

<!--
If you have identified a specific bug in the orchestrion codebase, please describe it here. If the
issue requires debugging or investigation of your specific private setup, that is best accomplished via support. Please contact Datadog [support](http://docs.datadoghq.com/help/).
-->

**Version of orchestrion**
<!-- If you aren't using the [latest version](https://github.com/DataDog/orchestrion/releases) of orchestrion, try upgrading first to see if your issue has already been resolved -->
1.0.4

**Describe what happened:**

Trying to instrument the `gin` tests using Orchestrion + CI Visibility throws build Id mismatch errors.

@RomainMuller discovered that the gin tests were running with `-covermode` and `-coverprofile` arguments without setting `-coverpkg`. If we pass the `-coverpkg` attribute then stops failing.

Gin makefile go test command: https://github.com/gin-gonic/gin/blob/master/Makefile#L15

 Log file:
[orchestrion-logs-ubuntu-latest-1.23- (3).zip](https://github.com/user-attachments/files/18660943/orchestrion-logs-ubuntu-latest-1.23-.3.zip)

Stdout:
```log
for d in github.com/gin-gonic/gin github.com/gin-gonic/gin/binding github.com/gin-gonic/gin/render; do \
	orchestrion go test  -v -covermode=count -coverprofile=profile.out $d > tmp.out; \
	cat tmp.out; \
	if grep -q "^--- FAIL" tmp.out; then \
		rm tmp.out; \
		exit 1; \
	elif grep -q "build failed" tmp.out; then \
		rm tmp.out; \
		exit 1; \
	elif grep -q "setup failed" tmp.out; then \
		rm tmp.out; \
		exit 1; \
	fi; \
	if [ -f profile.out ]; then \
		cat profile.out | grep -v "mode:" >> coverage.out; \
		rm profile.out; \
	fi; \
done
# os
resolving woven dependency on gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/emitter/ossec: -: # gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo
sending never-build-twice request: mismatched build ID for "gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo": "St17Dga3rWo_Bo1gn00A/St17Dga3rWo_Bo1gn00A" != "gDaWZDlp8ITA_dZ1NXih/gDaWZDlp8ITA_dZ1NXih"
FAIL	github.com/gin-gonic/gin [build failed]
FAIL
make: *** [Makefile:14: test] Error 1
Error: Process completed with exit code 2.
```

**Describe what you expected:**

Normal test pass

**Steps to reproduce the issue:**
<!-- If you can include a minimal code sample to reproduce the issue it can significantly reduce resolution time -->

1. Fork the gin repo: https://github.com/gin-gonic/gin
2. Modify the `Run Tests` step in https://github.com/gin-gonic/gin/blob/master/.github/workflows/gin.yml with:

```yaml
      - name: Run Tests
        run: |
          go install github.com/DataDog/orchestrion@latest
          make test
        env:
          DD_API_KEY: ${{ secrets.DD_API_KEY }}
```
3. Modify the Makefile Go command with:
```makefile
GO ?= orchestrion go
```
4. Run the github action

**Additional environment details (Version of Go, Operating System, etc.):**

--- a/internal/goflags/flags.go
+++ b/internal/goflags/flags.go
@@ -243,11 +243,11 @@ func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalA
 
 		for idx, pattern := range strings.Split(val, ",") {
 			if idx > 0 {
-				newValBuf.WriteByte(',')
+				_ = newValBuf.WriteByte(',')
 			}
 			if !strings.HasPrefix(pattern, "./") && !strings.HasPrefix(pattern, ".\\") {
 				// If the pattern is not relative, so we're good.
-				newValBuf.WriteString(pattern)
+				_, _ = newValBuf.WriteString(pattern)
 				continue
 			}
 
@@ -277,9 +277,9 @@ func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalA
 				}
 
 				if idx > 0 {
-					newValBuf.WriteByte(',')
+					_ = newValBuf.WriteByte(',')
 				}
-				newValBuf.WriteString(pkg.PkgPath)
+				_, _ = newValBuf.WriteString(pkg.PkgPath)
 			}
 		}
 

--- a/internal/goflags/flags.go
+++ b/internal/goflags/flags.go
@@ -219,7 +219,71 @@ func ParseCommandFlags(ctx context.Context, wd string, args []string) (CommandFl
 // inferCoverpkg will add the necessary `-coverpkg` argument if the `-cover` flags is present and `-coverpkg` is not, as
 // otherwise, sub-commands triggered with these flags will not apply coverage to the intended packages.
 func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalArgs []string) error {
-	if _, hasCoverpkg := f.Long["-coverpkg"]; hasCoverpkg {
+	log := zerolog.Ctx(ctx)
+
+	// Make sure we satisfy the same build constraints; but don't run -toolexec
+	childBuildFlags := append(f.Slice(), "-toolexec=")
+	childBuildLogf := func(format string, args ...any) {
+		log.Trace().Str("operation", "packages.Load").Msgf(format, args...)
+	}
+
+	if val, hasCoverpkg := f.Long["-coverpkg"]; hasCoverpkg {
+		if val == "" {
+			// Blank specified, not trying to expand it...
+			return nil
+		}
+
+		// We have patterns, we need to make sure they are expressed in absolute terms.
+		var newValBuf strings.Builder
+		newValBuf.Grow(len(val))
+
+		for idx, pattern := range strings.Split(val, ",") {
+			if idx > 0 {
+				newValBuf.WriteByte(',')
+			}
+			if !strings.HasPrefix(pattern, "./") && !strings.HasPrefix(pattern, ".\\") {
+				// If the pattern is not relative, so we're good.
+				newValBuf.WriteString(pattern)
+				continue
+			}
+
+			log.Debug().
+				Str("-coverpkg.entry", pattern).
+				Msg("Resolving relative -coverpkg entry")
+			pkgs, err := packages.Load(&packages.Config{
+				Mode:       packages.NeedName,
+				Dir:        wd,
+				BuildFlags: childBuildFlags,
+				Logf:       childBuildLogf,
+			}, pattern)
+			if err != nil {
+				return fmt.Errorf("resolving -coverpkg entry %q: %w", pattern, err)
+			}
+			for idx, pkg := range pkgs {
+				if len(pkg.Errors) != 0 {
+					var err error
+					for _, pkgErr := range pkg.Errors {
+						err = errors.Join(err, pkgErr)
+					}
+					log.Warn().
+						Err(err).
+						Str("pkg.ID", pkg.ID).
+						Str("-coverpkg.entry", pattern).
+						Msg("Error when resolving -coverpkg entry")
+				}
+
+				if idx > 0 {
+					newValBuf.WriteByte(',')
+				}
+				newValBuf.WriteString(pkg.PkgPath)
+			}
+		}
+
+		newVal := newValBuf.String()
+		f.Long["-coverpkg"] = newVal
+		log.Debug().
+			Str("-coverpkg", newVal).
+			Msg("Finalized -coverpkg value")
 		return nil
 	}
 
@@ -232,13 +296,12 @@ func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalA
 		return nil
 	}
 
-	log := zerolog.Ctx(ctx)
 	pkgs, err := packages.Load(
 		&packages.Config{
 			Mode:       packages.NeedName,
 			Dir:        wd,
-			BuildFlags: append(f.Slice(), "-toolexec"), // Make sure we satisfy the same build constraints; but don't run -toolexec
-			Logf:       func(format string, args ...any) { log.Trace().Str("operation", "packages.Load").Msgf(format, args...) },
+			BuildFlags: childBuildFlags,
+			Logf:       childBuildLogf,
 		},
 		positionalArgs...,
 	)

--- a/internal/goflags/flags.go
+++ b/internal/goflags/flags.go
@@ -216,8 +216,12 @@ func ParseCommandFlags(ctx context.Context, wd string, args []string) (CommandFl
 	return flags, nil
 }
 
-// inferCoverpkg will add the necessary `-coverpkg` argument if the `-cover` flags is present and `-coverpkg` is not, as
-// otherwise, sub-commands triggered with these flags will not apply coverage to the intended packages.
+// inferCoverpkg will add the necessary `-coverpkg` argument if the `-cover` flags is present and
+// `-coverpkg` is not, as otherwise, sub-commands triggered with these flags will not apply coverage
+// to the intended packages.
+// If `-coverpkg` is present, it will expand any relative paths (recognized by a `./` prefix) into
+// absolute package names, so that child builds do not interpret these relative to a different
+// package root.
 func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalArgs []string) error {
 	log := zerolog.Ctx(ctx)
 

--- a/internal/goflags/flags.go
+++ b/internal/goflags/flags.go
@@ -226,7 +226,7 @@ func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalA
 	_, isCover := f.Short["-cover"]
 	if !isCover {
 		// -covermode implies -cover
-		_, isCover = f.Short["-covermode"]
+		_, isCover = f.Long["-covermode"]
 	}
 	if !isCover {
 		return nil

--- a/internal/goflags/flags.go
+++ b/internal/goflags/flags.go
@@ -222,7 +222,13 @@ func (f *CommandFlags) inferCoverpkg(ctx context.Context, wd string, positionalA
 	if _, hasCoverpkg := f.Long["-coverpkg"]; hasCoverpkg {
 		return nil
 	}
-	if _, isCover := f.Short["-cover"]; !isCover {
+
+	_, isCover := f.Short["-cover"]
+	if !isCover {
+		// -covermode implies -cover
+		_, isCover = f.Short["-covermode"]
+	}
+	if !isCover {
 		return nil
 	}
 

--- a/internal/goflags/flags_test.go
+++ b/internal/goflags/flags_test.go
@@ -98,6 +98,13 @@ func TestParse(t *testing.T) {
 				Short: map[string]struct{}{"-cover": {}},
 			},
 		},
+		"covermode": {
+			flags: []string{"run", "-covermode=count"},
+			expected: CommandFlags{
+				Long:  map[string]string{"-covermode": "count", "-coverpkg": "github.com/DataDog/orchestrion/internal/goflags"},
+				Short: map[string]struct{}{},
+			},
+		},
 		"cover-with-coverpkg": {
 			flags:   []string{"run", "-cover", "-covermode=atomic", "--", "-some.go"},
 			goflags: "-coverpkg=std",

--- a/internal/goflags/flags_test.go
+++ b/internal/goflags/flags_test.go
@@ -102,7 +102,7 @@ func TestParse(t *testing.T) {
 			flags: []string{"run", "-covermode=count"},
 			expected: CommandFlags{
 				Long:  map[string]string{"-covermode": "count", "-coverpkg": "github.com/DataDog/orchestrion/internal/goflags"},
-				Short: map[string]struct{}{},
+				Short: nil,
 			},
 		},
 		"cover-with-coverpkg": {

--- a/internal/goflags/flags_test.go
+++ b/internal/goflags/flags_test.go
@@ -107,9 +107,9 @@ func TestParse(t *testing.T) {
 		},
 		"cover-with-coverpkg": {
 			flags:   []string{"run", "-cover", "-covermode=atomic", "--", "-some.go"},
-			goflags: "-coverpkg=std",
+			goflags: "-coverpkg=std,./...",
 			expected: CommandFlags{
-				Long:  map[string]string{"-covermode": "atomic", "-coverpkg": "std"},
+				Long:  map[string]string{"-covermode": "atomic", "-coverpkg": "std,github.com/DataDog/orchestrion/internal/goflags,github.com/DataDog/orchestrion/internal/goflags/quoted"},
 				Short: map[string]struct{}{"-cover": {}},
 			},
 		},

--- a/internal/pin/pin.go
+++ b/internal/pin/pin.go
@@ -34,9 +34,10 @@ const (
 )
 
 type Options struct {
-	// Writer is the writer to send output of the command to.
+	// Writer is the writer to send output of the command to. Defaults to
+	// [os.Stdout].
 	Writer io.Writer
-	// ErrWriter is the writer to send error messages to.
+	// ErrWriter is the writer to send error messages to. Defaults to [os.Stderr].
 	ErrWriter io.Writer
 
 	// Validate checks the contents of all [orchestrionDotYML] files encountered
@@ -55,6 +56,14 @@ type Options struct {
 // PinOrchestrion applies or update the orchestrion pin file in the current
 // working directory, according to the supplied [Options].
 func PinOrchestrion(ctx context.Context, opts Options) error {
+	// Ensure we have an [Options.Writer] and [Options.ErrWriter] set.
+	if opts.Writer == nil {
+		opts.Writer = os.Stdout
+	}
+	if opts.ErrWriter == nil {
+		opts.ErrWriter = os.Stderr
+	}
+
 	log := zerolog.Ctx(ctx)
 
 	goMod, err := goenv.GOMOD("")

